File: potato/src/path/rotation.py
from math import pow, fabs, sqrt

from src.location import Coordinates
from src.path.path import Path
from src.path.angle import Angle

class Rotation(Path):
    def __init__(self, x: float, y:float, start_theta: Angle, end_theta: Angle, max_speed: float = 2, max_acceleration: float = 2) -> None:
        super().__init__(going_backwards=False)
        self._x = x
        self._y = y
        self._start_theta = start_theta
        self._end_theta = end_theta
        self._max_speed = max_speed
        self._max_acceleration = max_acceleration

        self._angle_for_acceleration = 0.0
        self._angle_with_constant_speed = 0.0

        self._direction = -1 if (self._end_theta - self._start_theta).to_float() < 0 else 1

        acceleration_duration = self._max_speed / self._max_acceleration
        acceleration_angle = 0.5 * self._max_acceleration * pow(acceleration_duration, 2)
        total_angle = fabs((self._end_theta - self._start_theta).to_float())

        if 2 * acceleration_angle >total_angle:
            # Small rotation, we don't have time to reach max speed
            acceleration_angle = total_angle / 2
            acceleration_duration = sqrt(2 * acceleration_angle / self._max_acceleration)

            self._angle_for_acceleration = acceleration_angle
            self._angle_with_constant_speed = 0
            self._end_acceleration_time = acceleration_duration
            self._start_decceleration_time = acceleration_duration
            self._expected_duration = 2 * acceleration_duration
        else:
            # Big rotation. We will reach max speed
            self._angle_for_acceleration = acceleration_angle
            self._angle_with_constant_speed = total_angle - 2 * acceleration_angle
            self._end_acceleration_time = acceleration_duration
            self._start_decceleration_time = self._end_acceleration_time + self._angle_with_constant_speed / self._max_speed
            self._expected_duration = self._start_decceleration_time + acceleration_duration

    def expected_position(self, time: float) -> Coordinates:
        ellapsed_time = time - self._start_time

        if ellapsed_time < 0:
            return Coordinates(self._x, self._y, self._start_theta)
        elif 0 < ellapsed_time < self._end_acceleration_time:
            expected_theta = self._start_theta + Angle(0.5* self._direction * self._max_acceleration * pow(ellapsed_time, 2))
            return Coordinates(self._x, self._y, expected_theta)
        elif self._end_acceleration_time < ellapsed_time < self._start_decceleration_time:
            expected_theta = self._start_theta + Angle(self._direction *(self._angle_for_acceleration + (ellapsed_time-self._end_acceleration_time) * self._max_speed))
            return Coordinates(self._x, self._y, expected_theta)
        elif self._start_decceleration_time < ellapsed_time < self._expected_duration:
            expected_theta = self._end_theta - Angle(0.5*self._direction*self._max_acceleration*pow((self._expected_duration-ellapsed_time), 2))
            return Coordinates(self._x, self._y, expected_theta)
        else:
            return Coordinates(self._x, self._y, self._end_theta)

File: arduino/Stepper-motors/src/helpers/path/rotation.cpp
#include <Arduino.h>
#include <math.h>

#include "../maths_helpers.h"
#include "rotation.h"

Rotation::Rotation(double x, double y, Angle startTheta, Angle endTheta, double maxSpeed, double maxAcceleration, bool precision):Path()
{
  mX = x;
  mY = y;
  mStartTheta = startTheta;
  mEndTheta = endTheta;
  mMaxSpeed = maxSpeed;
  mMaxAcceleration = maxAcceleration;
  mPrecision = precision;

  if((mEndTheta-mStartTheta).toDouble()<0)
    mDirection = A_TRIGO_DIRECTION;
  else
    mDirection = TRIGO_DIRECTION;

  double accelerationDuration = mMaxSpeed / mMaxAcceleration;
  double accelerationAngle = 0.5*mMaxAcceleration*pow(accelerationDuration, 2);
  double totalAngle = fabs((mEndTheta-mStartTheta).toDouble());

  Serial.print("accelerationDuration ");
  Serial.print(accelerationDuration);
  Serial.print(" accelerationAngle ");
  Serial.print(accelerationAngle);
  Serial.print(" total angle ");
  Serial.println(totalAngle);

  if(2* accelerationAngle >totalAngle)
  {
    // Small rotation, we don't have time to reach max speed
    accelerationAngle = totalAngle / 2;
    accelerationDuration = sqrt(2*accelerationAngle/mMaxAcceleration);

    mAngleForAcceleration = accelerationAngle;
    mAngleWithConstantSpeed = 0;
    mEndAccelerationTime = accelerationDuration * SEC_TO_MICROS_MULTIPLICATOR;
    mStartDeccelerationTime = mEndAccelerationTime;
    mExpectedDuration = mStartDeccelerationTime + accelerationDuration * SEC_TO_MICROS_MULTIPLICATOR;
    Serial.print("Small rotation");
    Serial.print("expectedDuration ");
    Serial.println(mExpectedDuration);
  }
  else
  {
    // Big rotation. We will reach max speed
    mAngleForAcceleration = accelerationAngle;
    mAngleWithConstantSpeed = totalAngle-2 * mAngleForAcceleration;
    mEndAccelerationTime = accelerationDuration * SEC_TO_MICROS_MULTIPLICATOR;
    mStartDeccelerationTime = mEndAccelerationTime + SEC_TO_MICROS_MULTIPLICATOR * mAngleWithConstantSpeed / mMaxSpeed;
    mExpectedDuration = mStartDeccelerationTime + accelerationDuration * SEC_TO_MICROS_MULTIPLICATOR;
    Serial.print("Big rotation");
    Serial.print("expectedDuration ");
    Serial.println(mExpectedDuration);
  }
}

Rotation::~Rotation(){}

String Rotation::debugString() 
{
  String startTheta = String(mStartTheta.toDouble(), 4);
  String endTheta = String(mEndTheta.toDouble(), 4);
  return String("Rotation from " + startTheta + " to " + endTheta);
}

double Rotation::positionError(double x, double y, Angle theta, long time)
{
  double vector[2] = {cos(theta.toDouble()), sin(theta.toDouble())};
  double projectedX = (mX-x)*vector[X]+(mY-y)*vector[Y];
  return projectedX;
}

double Rotation::rotationError(double x, double y, Angle theta, long time)
{
  Serial.print("time ");
  Serial.print(time);
  Serial.print(" startTime ");
  Serial.print(mStartTime);
  Serial.print(" ");

  long ellapsedTime = time - mStartTime;
  if(ellapsedTime < 0)
  {
    Serial.print(" 1 ");
    return (mStartTheta-theta).toDouble();
  }
  else if(0 < ellapsedTime && ellapsedTime < mEndAccelerationTime)
  {
    Serial.print(" 2 ");
    Angle expectedTheta = mStartTheta + 0.5* mDirection * mMaxAcceleration *pow((ellapsedTime) * MICROS_TO_SEC_MULTIPLICATOR, 2);
    return (expectedTheta-theta).toDouble();
  }
  else if(mEndAccelerationTime < ellapsedTime && ellapsedTime < mStartDeccelerationTime)
  {
    Serial.print(" 3 ");
    Angle expectedTheta = mStartTheta + mDirection *(mAngleForAcceleration + (ellapsedTime-mEndAccelerationTime)*MICROS_TO_SEC_MULTIPLICATOR * mMaxSpeed);
    return (expectedTheta-theta).toDouble();
  }
  else if(mStartDeccelerationTime < ellapsedTime && ellapsedTime < mExpectedDuration)
  {
    Serial.print(" 4 ");
    Angle expectedTheta = mEndTheta - 0.5*mDirection*mMaxAcceleration*pow((mExpectedDuration-ellapsedTime)* MICROS_TO_SEC_MULTIPLICATOR, 2);
    return (expectedTheta-theta).toDouble();
  }
  else
  {
    Serial.print(" 5 ");
    return (mEndTheta-theta).toDouble();
  }
  Serial.println(" ");
}

bool Rotation::needsPrecision() const {
  return mPrecision;
}

